### PR TITLE
Allow to specify custom tags to decide the assignment of servers

### PR DESCRIPTION
**Pull request by zuston**
### What changes were proposed in this pull request?
Allow to specify custom tags to decide data placement.
Changelog
1. Introduce the conf of `rss.server.tags` for shuffle server
2. Introduce the conf of `rss.client.assignment.tags` for client to choose the data placement

### Why are the changes needed?
Sometimes, we hope the specified spark/mr job's shuffle data can be stored the specified group of shuffle servers, maybe due to DC location and so on.

### Does this PR introduce _any_ user-facing change?
Yes.

### How was this patch tested?
UTs.
<!--

### CODE REVIEW

**Comment by zuston** [comment]
@jerqi Could u help review?

**Comment by jerqi** [comment]
You should add some integration test cases.

**Comment by zuston** [comment]
> You should add some integration test cases.

Yes, i’m doing it. Sorry not to add it firstly.

File: server/src/main/java/org/apache/uniffle/server/ShuffleServer.java [hunk 1]
```diff
@@ -262,7 +274,7 @@ public StorageManager getStorageManager() {
   }
 
   public Set<String> getTags() {
-    return tags;
+    return Collections.unmodifiableSet(tags);
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/ShuffleServer.java:277]: Why do we need `Collections.unmodifiableSet`?

Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServer.java:277]: I think this method is to return read -only tags.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/v2/app/RssMRAppMaster.java [hunk 2]
```diff
@@ -115,11 +117,19 @@ public static void main(String[] args) {
       LOG.info("Registering coordinators {}", coordinators);
       client.registerCoordinators(coordinators);
 
+      // Get the configured data placement tags and it will also add default shuffle version tag.
+      Set<String> dataPlacementTags = new HashSet<>();
+      String tagsRaw = conf.get(RssMRConfig.RSS_CLIENT_DATA_PLACEMENT_TAGS,
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/v2/app/RssMRAppMaster.java:122]: rawTags?

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssSparkShuffleUtils.java [hunk 3]
```diff
@@ -123,4 +127,15 @@ public static Configuration getRemoteStorageHadoopConf(
     }
     return readerHadoopConf;
   }
+
+  public static Set<String> getDataPlacementTags(SparkConf sparkConf) {
+    Set<String> placementTags = new HashSet<>();
+    String tagsRaw = sparkConf.get(RssSparkConfig.RSS_CLIENT_DATA_PLACEMENT_TAGS,
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/RssSparkShuffleUtils.java:133]: ditto?

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssSparkShuffleUtils.java [hunk 4]
```diff
@@ -123,4 +127,15 @@ public static Configuration getRemoteStorageHadoopConf(
     }
     return readerHadoopConf;
   }
+
+  public static Set<String> getDataPlacementTags(SparkConf sparkConf) {
+    Set<String> placementTags = new HashSet<>();
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/RssSparkShuffleUtils.java:132]: Could we give the variable a better name?

File: client/src/main/java/org/apache/uniffle/client/util/RssClientConfig.java [hunk 5]
```diff
@@ -56,6 +58,9 @@ public class RssClientConfig {
   // When the size of read buffer reaches the half of JVM region (i.e., 32m),
   // it will incur humongous allocation, so we set it to 14m.
   public static String RSS_CLIENT_READ_BUFFER_SIZE_DEFAULT_VALUE = "14m";
+  // The tags specified by rss client to determine shuffle data placement.
+  public static String RSS_CLIENT_DATA_PLACEMENT_TAGS = "rss.client.data.placement.tags";
```
Comment (jerqi) [inline client/src/main/java/org/apache/uniffle/client/util/RssClientConfig.java:62]: Could we give it a better name ?
The reason is as below:
https://github.com/apache/incubator-uniffle/pull/30#issuecomment-1175014084

File: integration-test/common/src/test/java/org/apache/uniffle/test/AssignmentWithTagsTest.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        }
+
+        for (int i = 0; i < 2; i ++) {
+            createAndStartShuffleServerWithTags(Sets.newHashSet());
+        }
+
+        for (int i = 0; i < 2; i++) {
+            createAndStartShuffleServerWithTags(Sets.newHashSet("fixed"));
+        }
+
```
Comment (jerqi) [inline integration-test/common/src/test/java/org/apache/uniffle/test/AssignmentWithTagsTest.java:129]: Could we add another server with another tag?

File: client/src/main/java/org/apache/uniffle/client/util/RssClientConfig.java [hunk 7]
```diff
@@ -56,6 +58,9 @@ public class RssClientConfig {
...
+  public static String RSS_CLIENT_DATA_PLACEMENT_TAGS_DEFAULT_VALUES = Constants.SHUFFLE_SERVER_VERSION;
```
Comment (jerqi) [inline client/src/main/java/org/apache/uniffle/client/util/RssClientConfig.java:63]: A shuffle server can have multiple tags. The version tag shouldn't be overrided the configurable tags.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/RssSparkShuffleUtils.java (hunk 4 above)
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/RssSparkShuffleUtils.java:132]: server assignment tag?

File: client/src/main/java/org/apache/uniffle/client/util/RssClientConfig.java (hunk 5 above)
Comment (zuston) [reply client/src/main/java/org/apache/uniffle/client/util/RssClientConfig.java:62]: rss.client.shuffle-server.assignment.tags ?

Comment (jerqi) [reply client/src/main/java/org/apache/uniffle/client/util/RssClientConfig.java:62]: `rss.client.assignment.tags` will be better.

**Review by jerqi — APPROVED** [review]
LGTM, thanks for your contribution.

**Comment by jerqi** [comment]
Would you like to add some docs about this feature in another pr?

**Comment by zuston** [comment]
> Would you like to add some docs about this feature in another pr? Because this feature brings some user-facing changes, we need some docs.

Glad to do. I will open another PR to finish.

**Pull request merged**